### PR TITLE
 Port deviceManager to windows container manager to enable GPU access

### PR DESCRIPTION
**Pull request by aarnaud**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Port the devicemanager to Windows node to allow device plugins.

An important device plugin (not part of this PR) is the directx device plugin for GPU access: https://github.com/aarnaud/k8s-directx-device-plugin

This is required for machine learning, high-performance computing and computer graphics under Windows using the GPU.

References:
https://github.com/kubernetes/kubernetes/pull/80917
https://github.com/kubernetes/kubernetes/pull/93263
https://github.com/kubernetes/kubernetes/issues/93262
https://github.com/kubernetes/kubernetes/issues/88554

Fix issue:
https://github.com/kubernetes/kubernetes/issues/88157

**Does this PR introduce a user-facing change?**:

```release-note
Port the devicemanager to Windows node to allow device plugins like directx
```

### CODE REVIEW

**Comment by aarnaud** [comment]
Hi @matthyx and @michmike can you add ok-to-test label on this PR, with windows test fixed now.

**Comment by thomacos** [comment]
/approve
Moral support.

**Comment by aarnaud** [comment]
Hi @jsturtevant @liggitt 

I add again the constant file for windows to avoid this crash, and test it on windows server 2019 and all test passed.
Origin of this issue: https://github.com/kubernetes/kubernetes/pull/80917#discussion_r450490699

**Comment by aarnaud** [comment]
/milestone v1.19

@michmike can you add this in the v1.19 like https://github.com/kubernetes/kubernetes/pull/80917

now the windows e2e test work and pass, we need to add test for device plugin on windows but I'm not comfortable with e2e test system on this project.

File: staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go
Comment (liggitt) [inline staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go:29]: OS-specific constants are pretty unusual ... it means code that references this constant but is not running on the target OS will get the wrong value

**Comment by thomacos** [comment]
Hi @derekwaynecarr,

Can you approve this PR for allowing GPU access under Windows?

This PR is the same as https://github.com/kubernetes/kubernetes/pull/80917, which was already fully approved and merged, except for the following change:
A Windows-specific fix of bug https://github.com/kubernetes/kubernetes/issues/93262 requires setting the DevicePluginPath constant to a different value under Windows than under Linux https://github.com/kubernetes/kubernetes/pull/93285/files#diff-9f69ebb59ab7424eb7662a8f6046467bR29.
All tests are passing.

File: staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go
Comment (aarnaud) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go:29]: I understand but how can I change that. by using `goruntime.GOOS != "windows"` in  https://github.com/kubernetes/kubernetes/blob/5a529aa3a0dd3a050c5302329681e871ef6c162e/pkg/kubelet/cm/devicemanager/manager.go#L134

https://github.com/kubernetes/kubernetes/blob/6aa4c7eca83b74076ae24afde5edff7d932b854e/cmd/kubeadm/app/util/runtime/runtime.go#L62

Comment (thomacos) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go:29]: Device plugin paths must be absolute path, see https://github.com/kubernetes/kubernetes/blob/5a529aa3a0dd3a050c5302329681e871ef6c162e/pkg/kubelet/cm/devicemanager/manager.go#L134.

However, the definition of "absolute path" is different on Windows and Linux, see the different implementations of `filepath.IsAbs`, see https://golang.org/src/path/filepath/path_windows.go versus https://golang.org/src/path/filepath/path_unix.go.

Therefore, it is unavoidable that the DevicePluginPath is different under Windows and Linux.

Comment (liggitt) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go:29]: it looks like DevicePluginPath is only referenced by KubeletSocket, and KubeletSocket is only referenced by NewManagerImpl in-tree. If we needed a different default path per-platform, something like this would be clearer to me:

```diff
diff --git a/pkg/kubelet/cm/devicemanager/manager.go b/pkg/kubelet/cm/devicemanager/manager.go
index 5d1925f9458..14dc3be1347 100644
--- a/pkg/kubelet/cm/devicemanager/manager.go
+++ b/pkg/kubelet/cm/devicemanager/manager.go
@@ -22,6 +22,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"runtime"
 	"sort"
 	"sync"
 	"time"
@@ -125,7 +126,11 @@ func (s *sourcesReadyStub) AllReady() bool          { return true }
 
 // NewManagerImpl creates a new manager.
 func NewManagerImpl(numaNodeInfo cputopology.NUMANodeInfo, topologyAffinityStore topologymanager.Store) (*ManagerImpl, error) {
-	return newManagerImpl(pluginapi.KubeletSocket, numaNodeInfo, topologyAffinityStore)
+	socketPath := pluginapi.KubeletSocket
+	if runtime.GOOS == "windows" {
+		socketPath = pluginapi.KubeletSocketWindows
+	}
+	return newManagerImpl(socketPath, numaNodeInfo, topologyAffinityStore)
 }
 
 func newManagerImpl(socketPath string, numaNodeInfo cputopology.NUMANodeInfo, topologyAffinityStore topologymanager.Store) (*ManagerImpl, error) {
```

Comment (thomacos) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go:29]: Yes, ideally all paths are the same under Windows and Unix.

However, a) the overall convention of Kubernetes using hard-wired absolute paths, together with b) the meaning of absolute paths being different on Unix and Windows unavoidably leads to the conclusion that we need different path definitions on Unix and Windows.

And in this case I don't see how this is critical anyway, since I don't see any realistic use-case to use `DevicePluginPath` in a cross-compiling situation as you are mentioning.

Comment (aarnaud) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go]: Thanks @liggitt, I changed my PR to add your feedback

File: staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go
Comment (michmike) [inline staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go:38]: will it always be on c:\ drive?

**Review by michmike — APPROVED** [review]
/lgtm
/approve

**Comment by michmike** [comment]
cc: @klueska can you please take a look

File: staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go
Comment (thomacos) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go:38]: The clean solution seems to be https://github.com/kubernetes/kubernetes/pull/91118.
But because the path was hardwired on Unix so far, I think doing the same for Windows should not block this PR.

Comment (aarnaud) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go:38]: It's seem they use systemDrive
```
$global:KubernetesPath = "$env:SystemDrive\k"
```
https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/kubeadm/scripts/PrepareNode.ps1

Comment (michmike) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go:38]: should you also use $env:SystemDrive in your PR instead of C:\ then?

Comment (thomacos) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go:38]: > should you also use $env:SystemDrive in your PR instead of C:\ then?

But then applying #91118 first seems cleaner to me.

Comment (aarnaud) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go]: > should you also use $env:SystemDrive in your PR instead of C:\ then?

yes, I just add it

https://github.com/kubernetes/kubernetes/pull/93285/commits/e03f8287c942709cb1ddeaf17f0ea6126caf2095#diff-b54ee8195220ea88e59b6eb423e83dd1R131

Comment (thomacos) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants.go]: > 
> 
> > should you also use $env:SystemDrive in your PR instead of C:\ then?
> 
> yes, I just add it
> 
> [e03f828#diff-b54ee8195220ea88e59b6eb423e83dd1R131](https://github.com/kubernetes/kubernetes/commit/e03f8287c942709cb1ddeaf17f0ea6126caf2095#diff-b54ee8195220ea88e59b6eb423e83dd1R131)

Yes, that's nice!
Then #91118 can be done separately.

**Review by michmike — APPROVED** [review]
/lgtm
/approve

**Comment by thomacos** [comment]
> /retest

It seems that testing is not fully deterministic.
A full retest risks that sometimes tests may fail after they had succeeded previously.
This can be avoided by only retesting the failed tests.

**Comment by thomacos** [comment]
> 
> 
> > /retest
> 
> It seems that testing is not fully deterministic.
> A full retest risks that sometimes tests may fail after they had succeeded previously.
> This can be avoided by only retesting the failed tests.

A, I just learned that /retest is exactly doing this, what I had done by hand previously.

**Comment by thomacos** [comment]
@liggitt Since you made #93263 approved by giving your approval, can you do the same for this PR?

I must admit I don't understand the process: Your approval of #93263 together with @michmike's approval seemed to have put it over the finish line, although you are not in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/OWNERS.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/93263 rolled back https://github.com/kubernetes/kubernetes/pull/80917 since it left the current windows CI in a non-functional state

I'll defer to sig-node / sig-windows on reintroduction of the changes, but given how far past freeze we are, and that there are no accompanying tests, I would recommend merging this for 1.20 (with test coverage as described in https://github.com/kubernetes/kubernetes/issues/88554, not as a followup)

> I must admit I don't understand the process: Your approval of #93263 together with @michmike's approval seemed to have put it over the finish line, although you are not in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/OWNERS.

approval is hierarchical, so approvers in https://github.com/kubernetes/kubernetes/blob/master/pkg/OWNERS and https://github.com/kubernetes/kubernetes/blob/master/OWNERS can approve PRs to subdirectories as well

**Comment by thomacos** [comment]
@liggitt Thanks for the infos! You live and learn.

May I ask you if you could approve this PR for merging into the master branch (which won't go into 1.19 due to Code Thaw).
And then we work on #88554 e2e test for 1.20 as you recommended.

Sorry for asking this again, but I'm afraid that this PR will get lost again: The previous version of this PR https://github.com/kubernetes/kubernetes/pull/80917 took nearly a year to be even recognized beyond sig-windows.

This PR is already approved by @michmike co-chair for sig-windows. All tests are passing. And this PR is exactly the same as https://github.com/kubernetes/kubernetes/pull/80917, which was fully discussed, reviewed, approved and merged, except for this trivial Windows-specific fix.

> #93263 rolled back #80917 since it left the current windows CI in a non-functional state
> 
> I'll defer to sig-node / sig-windows on reintroduction of the changes, but given how far past freeze we are, and that there are no accompanying tests, I would recommend merging this for 1.20 (with test coverage as described in #88554, not as a followup)
> 
> > I must admit I don't understand the process: Your approval of #93263 together with @michmike's approval seemed to have put it over the finish line, although you are not in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/OWNERS.
>
> approval is hierarchical, so approvers in https://github.com/kubernetes/kubernetes/blob/master/pkg/OWNERS and https://github.com/kubernetes/kubernetes/blob/master/OWNERS can approve PRs to subdirectories as well

**Comment by liggitt** [comment]
> May I ask you if you could approve this PR for merging into the master branch (which won't go into 1.19 due to Code Thaw).
> And then we work on #88554 e2e test for 1.20 as you recommended.

That seems like the wrong order to me. I would expect test coverage of functionality before we merge it.

**Comment by michmike** [comment]
@liggitt this PR was tested. it is just that no automated tests were created for coverage yet. We obviously want all PRs that come in to have automated tests, but that does not always happen

**Comment by thomacos** [comment]
> @liggitt this PR was tested. it is just that no automated tests were created for coverage yet. We obviously want all PRs that come in to have automated tests, but that does not always happen

@liggitt Yes, we are successfully using a custom build of Kubernetes to use the directx device plugin https://github.com/aarnaud/k8s-directx-device-plugin for GPU access. 
This is exactly the reason we are keen on getting this PR merged, so that we don't have to use a Kubernetes custom build anymore.

**Comment by liggitt** [comment]
In our experience, functionality that is only verified manually is not reliable or supportable ... having automated tests that verify a new component works in at least a basic way is a reasonable prerequisite for merging

From https://github.com/kubernetes/community/blob/master/contributors/guide/coding-conventions.md#testing-conventions:

> Significant features should come with integration (test/integration) and/or end-to-end (test/e2e) tests

**Comment by thomacos** [comment]
> In our experience, functionality that is only verified manually is not reliable or supportable ... having automated tests that verify a new component works in at least a basic way is a reasonable prerequisite for merging
> 
> From https://github.com/kubernetes/community/blob/master/contributors/guide/coding-conventions.md#testing-conventions:
> 
> > Significant features should come with integration (test/integration) and/or end-to-end (test/e2e) tests

I agree completely with the usefulness of tests for any kind of feature - definitely no doubt.
I fully understand your point of view.

But I also understand the reluctance of a voluntary contributor, after having already invested work in creating a PR, but being ignored by relevant approvers for a year, then finally approved, but then being required to jump through another new hoop.

By the way, I wouldn't call this PR a "significant feature", but a very trivial port of an existing feature (which has an e2e-test) from Unix to Windows. And while more e2e tests would be great of course, I also would say that we already have "basic ways" of automated tests of this PR, as evidently demonstrated by https://github.com/kubernetes/kubernetes/issues/93262.

**Comment by BenTheElder** [comment]
> But I also understand the reluctance of a voluntary contributor, after having already invested work in creating a PR, but being ignored by relevant approvers for a year, then finally approved, but then being required to jump through another new hoop.

liggitt is not one of these approvers / reviewers though, and the "approval" so far is not by someone who owns this code (or else we wouldn't still be discussing this).

Another PR that is a superset of your work could accomplish this, if we'd like someone else to deal with this request..

> By the way, I wouldn't call this PR a "significant feature", but a very trivial port of an existing feature (which has an e2e-test) from Unix to Windows. And while more e2e tests would be great of course, I also would say that we already have "basic ways" of automated tests of this PR, as evidently demonstrated by #93262.

AIUI we have an e2e test, but not one that works under Windows? (And therefore this PR?)

"Enable GPU access on another platform" is a pretty significant feature, imho.

Adding an e2e test is a pretty standard review comment, that certainly should have come up a year ago.

SIG Contributor experience is working on the inactive reviewer/approver/owner problem.
https://groups.google.com/d/msgid/kubernetes-dev/CAOmijWw6GrNZtHZAurnfSQrZv25PwdUZi___CW78SGL-E3-Shg%40mail.gmail.com?utm_medium=email&utm_source=footer

**Comment by liggitt** [comment]
> By the way, I wouldn't call this PR a "significant feature", but a very trivial port of an existing feature (which has an e2e-test) from Unix to Windows.

The environmental differences between windows and linux are significant enough that verifying the feature works specifically on windows seems important

> And while more e2e tests would be great of course, I also would say that we already have "basic ways" of automated tests of this PR, as evidently demonstrated by #93262.

The azure CI jobs now passing on this PR let us know this doesn't prevent kubelet startup, which is necessary but not sufficient. If I understand correctly, they don't exercise the functionality added by this PR.

**Comment by liggitt** [comment]
I'll also note that my top-level approval is not required to merge this:

> Needs approval from an approver in each of these files:
> * pkg/kubelet/cm/OWNERS
> * staging/src/k8s.io/kubelet/OWNERS

However, I would encourage those component owners to limit approval to things which have sufficient basic test coverage

**Comment by thomacos** [comment]
> > But I also understand the reluctance of a voluntary contributor, after having already invested work in creating a PR, but being ignored by relevant approvers for a year, then finally approved, but then being required to jump through another new hoop.
> 
> liggitt is not one of these approvers / reviewers though, and the "approval" so far is not by someone who owns this code (or else we wouldn't still be discussing this).

I was referring to the original https://github.com/kubernetes/kubernetes/pull/80917.

> Another PR that is a superset of your work could accomplish this, if we'd like someone else to deal with this request..

Minor detail, but it's not my code, I'm "just" interested into this features.
(We are currently working with a custom build of Kubernetes including this feature.)

> > By the way, I wouldn't call this PR a "significant feature", but a very trivial port of an existing feature (which has an e2e-test) from Unix to Windows. And while more e2e tests would be great of course, I also would say that we already have "basic ways" of automated tests of this PR, as evidently demonstrated by #93262.
> 
> AIUI we have an e2e test, but not one that works under Windows? (And therefore this PR?)

This PR is porting a feature from Unix to Windows.
There is a e2e test of this feature for Unix, but not for Windows yet, see https://github.com/kubernetes/kubernetes/issues/88554.

> "Enable GPU access on another platform" is a pretty significant feature, imho.

This PR is not providing GPU access.
This PR is only porting the existing device plugin API from Unix to Windows.
But the device plugin API is required for running the directx plugin, which is not part of this PR, but provides GPU access.
In other words, this PR is "only" the required basis for providing GPU access via a plugin.

> Adding an e2e test is a pretty standard review comment, that certainly should have come up a year ago.
> 
> SIG Contributor experience is working on the inactive reviewer/approver/owner problem.
> https://groups.google.com/d/msgid/kubernetes-dev/CAOmijWw6GrNZtHZAurnfSQrZv25PwdUZi___CW78SGL-E3-Shg%40mail.gmail.com?utm_medium=email&utm_source=footer

To clarify, I definitely don't want to blame anyone. Everyone has too much on its plate, plus unexpected problems we all are familiar with, causing unintended consequences here and there without anyone having done something wrong. 

And btw, thanks a lot to all of you for your great work on Kubernetes!

**Comment by thomacos** [comment]
> > By the way, I wouldn't call this PR a "significant feature", but a very trivial port of an existing feature (which has an e2e-test) from Unix to Windows.
> 
> The environmental differences between windows and linux are significant enough that verifying the feature works specifically on windows seems important
> 
> > And while more e2e tests would be great of course, I also would say that we already have "basic ways" of automated tests of this PR, as evidently demonstrated by #93262.
> 
> The azure CI jobs now passing on this PR let us know this doesn't prevent kubelet startup, which is necessary but not sufficient. If I understand correctly, they don't exercise the functionality added by this PR.

I also understand this point of view.

> I'll also note that my top-level approval is not required to merge this:
> 
> > Needs approval from an approver in each of these files:
> > 
> > * pkg/kubelet/cm/OWNERS
> > * staging/src/k8s.io/kubelet/OWNERS
> 
> However, I would encourage those component owners to limit approval to things which have sufficient basic test coverage

Frankly, since you showed your finger, I am just grabbing your hand, because such approvers you are mentioning were requested for the original PR a year ago, but never responded in any way. (Again, no criticism, just an explanation for my worry.)

The original PR was only merged because I found and triggered some other approver. But to my understanding this approver has not enough approval rights for this slightly modifed PR because of the trivial change of the device plugin path handling.

This is exactly the reason why I'm worried that when now "loosing" you, this PR will be buried forever... Or to express it emphatically, you are my last hope for this PR ...

**Comment by liggitt** [comment]
> This PR is porting a feature from Unix to Windows. There is a e2e test of this feature for Unix, but not for Windows yet, see #88554.

I'm not sure why the test was considered a follow-up and not a prereq. In the past, we have found notable differences between linux and windows in listener / socket / pipe handling (in addition to the normal suspects around path handling). A few examples:
* https://github.com/kubernetes/kubernetes/issues/78628#issuecomment-498239823 (looks like that was actually [referenced](https://github.com/kubernetes/kubernetes/pull/80917#issuecomment-586539665) in the review of the original version of this PR)
* kubectl port-forward - https://github.com/kubernetes/kubernetes/issues/78857
* CSI support on windows - https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190714-windows-csi-support.md#implementation-details

Given that history, and the fact that most contributors are not running windows node environments, having tests in place to ensure the ported functionality works and keeps working is very important.

> Frankly, since you showed your finger, I am just grabbing your hand, because such approvers you are mentioning were requested for the original PR a year ago, but never responded in any way. (Again, no criticism, just an explanation for my worry.)
>
> The original PR was only merged because I found and triggered some other approver. But to my understanding this approver has not enough approval rights for this slightly modifed PR because of the trivial change of the device plugin path handling.

My general recommendation in a situation like that is to request sig involvement via the normal sig meeting (https://github.com/kubernetes/community/tree/master/sig-windows#meetings), adding an item to the agenda, and requesting commitment from an approver for review time.

**Comment by thomacos** [comment]
It seems to me that in this case the risks are rather low, but indeed I don't know the history you are referring to.

Just to clarify: sig-windows was and is not the problem, but getting approval from relevant non-sig-windows approvers. 
Which is ironic for a PR which is purely Windows related.

In any case, thanks for your time and valuable input!

**Comment by thomacos** [comment]
(I fixed my language in my previous post, sorry for that. I wanted to say: " ... since you gave me your hand I am grabbing your whole arm ..." English is not my primary language.)

**Comment by michmike** [comment]
everyone, given where we are 1.19 and the fact that we trying to wrap things up, i am pushing this PR to v1.20. @aarnaud and @thomacos thank you so much for your effort so far. as we are discussing in https://github.com/kubernetes/kubernetes/issues/88554, we need automated tests. if we provide those, we should be able to get this into 1.20.
/milestone v1.20

File: staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go
Comment (ddebroy) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go]: > OS-specific constants are pretty unusual ... it means code that references this constant but is not running on the target OS will get the wrong value

My understanding here was: only the builds targeting Windows will reference the constants in `constants_windows.go` while Linux builds will reference the corresponding constants in `constants_linux.go` / `constants_unix.go` based on: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants_windows.go and  https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta1/defaults_windows.go  as well as in Go libraries: https://github.com/golang/go/blob/master/src/os/path_windows.go#L8

@liggitt can you explain a bit why/how this does not work. I assume the compiler will fail the build if a targeted platform does not have all the constants defined that the platform agnostic code is referring to?

**Comment by thomacos** [comment]
> everyone, given where we are 1.19 and the fact that we trying to wrap things up, i am pushing this PR to v1.20. @aarnaud and @thomacos thank you so much for your effort so far. as we are discussing in #88554, we need automated tests. if we provide those, we should be able to get this into 1.20.
> /milestone v1.20

Makes sense.

> please attend the next sig-windows community meeting if you would like to discuss this further.
> https://github.com/kubernetes/community/tree/master/sig-windows#meetings

I'll try to be there.

File: staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go
Comment (BenTheElder) [reply staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1/constants_windows.go]: > OS-specific constants are pretty unusual ... it means code that references this constant but is not running on the target OS will get the wrong value

@ddebroy I think @liggitt's concern here is roughly that because we use a single name across platforms for this constant, it may not be obvious when we wind up with code that depends on properties of a specific platform's constant value.

As opposed to having distinct constants for each platform.

If we already have platform specific implementations referencing the constant, why not reference a platform specific constant instead ...?

 > it looks like DevicePluginPath is only referenced by KubeletSocket, and KubeletSocket is only referenced by NewManagerImpl in-tree. If we needed a different default path per-platform, something like this would be clearer to me:

This comment outlines this before showing a specific proposed implementation ...

**Comment by liggitt** [comment]
this will be moved back into the milestone as part of the [phased reopen for 1.20 merges](https://docs.google.com/document/d/1yI7o7R2bOB7QtkHUFMHMb5XObiJmzIAnJPuxhdAX-CA/edit)

**Comment by immuzz** [comment]
@aarnaud can you rebase this?

**Comment by aarnaud** [comment]
> @aarnaud can you rebase this?

@immuzz done

**Comment by aarnaud** [comment]
> Please fix the lint errors in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93285/pull-kubernetes-verify/1315492267875635200

Oh yes I miss that,
it fixed. waiting for test result

**Comment by thomacos** [comment]
The current version of that PR is not working with the latest Kubernetes version, causing a null reference exception.
My colleague is planning to send a fix soon.

**Comment by aarnaud** [comment]
do you have the stacktrace ? and the reproductive case ? I can check and fix the code.

**Comment by b3atlesfan** [comment]
> do you have the stacktrace ? and the reproductive case ? I can check and fix the code.

Here you go.
https://github.com/aarnaud/kubernetes/pull/1#issuecomment-724040157

**Comment by aarnaud** [comment]
Thanks @b3atlesfan for the fix, I added in my PR.

**Comment by b3atlesfan** [comment]
> Thanks @b3atlesfan for the fix, I added in my PR.

Glad, I could help.

**Comment by thomacos** [comment]
@aarnaud Can you merge these two changes:

- https://github.com/kubernetes/kubernetes/pull/93948/commits/683bffacf68de8d1e8882a71c3a5e9525141c881
- https://github.com/kubernetes/kubernetes/pull/93948/commits/d72bd491ead24be0b61aec3e9fad04b349487679

This fixes an nil reference crash, see
- https://github.com/kubernetes/kubernetes/pull/93948#issuecomment-726575046
- https://github.com/kubernetes/kubernetes/pull/93948#issuecomment-726747240

After this I think this PR is fully ready to be merged into Kubernetes, since we have a e2e test https://github.com/kubernetes/kubernetes/pull/93948 which verifies both registering a device plugin and a Windows container successfully requesting and getting a device.

**Comment by aarnaud** [comment]
Hi @thomacos I rebase this PR on master and add  the fix for nil reference crash

**Comment by thomacos** [comment]
@michmike
Can you reintroduce this PR as suggested by @liggitt in https://github.com/kubernetes/kubernetes/pull/93285#issuecomment-662192622? 

In the past, the reason for the holdup was a missing e2e test. Now we have it: https://github.com/kubernetes/kubernetes/pull/93948. It successfully deploys the Windows DirectX device plugin, creates a Windows test container which requests and retrieves a DirectX device successfully: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/93948/pull-kubernetes-e2e-aks-engine-windows-gpu/1327233355753721856

In other words, all reasons for the holdups of this PR should be resolved now.
Ideally we get it into 1.20.

**Review by michmike — APPROVED** [review]
/approve
/lgtm

**Comment by michmike** [comment]
@klueska you have reviewed the earlier version of the PR (https://github.com/kubernetes/kubernetes/pull/80917). could you please review this again? E2e tests are passing as per https://github.com/kubernetes/kubernetes/issues/88554 and we should merge those soon as well

**Comment by marosset** [comment]
The changes look good to me. 
Let me sync with @michmike to see if we will file an exception for 1.20 or wait until 1.21.

**Comment by erismaster** [comment]
Hi 👋🏽 I'm from the Bug Triage team. We've crossed the Code Freeze for 1.20 release on 12th November. As this PR is tagged with 1.20, I'm sending a final reminder to either move the milestone to 1.21 or clear the milestone.

**Comment by jeremyrickard** [comment]
Hey @michmike , 1.20 Release Lead :wave:,

As @liggitt pointed out, we're well past code freeze at this point and we'll need to bump this to 1.21, unless this is *truly* release blocking? Looking at this, it does not appear to be. We're really only considering exception requests for feature things on a very very limited basis (i.e. release blocking). For now, let's bump this to v1.21. Feel free to reach out and discuss this if you feel otherwise.


/milestone v1.21

**Comment by thomacos** [comment]
> Hey @michmike , 1.20 Release Lead 👋,
> 
> As @liggitt pointed out, we're well past code freeze at this point and we'll need to bump this to 1.21, unless this is _truly_ release blocking? Looking at this, it does not appear to be. We're really only considering exception requests for feature things on a very very limited basis (i.e. release blocking). For now, let's bump this to v1.21. Feel free to reach out and discuss this if you feel otherwise.
> 
> /milestone v1.21

Hi @jeremyrickard, 

I would argue that this PR fits well into an "[Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md)" for merging into v1.20:

- It has "low risk" - in fact it is not even an real enhancement, but a trivial port of a feature from Linux to Windows.
- It doesn't "risk other areas of the code" and is "well contained and tested" (by a full e2e test).
- The "length of exception needed" is "on the order of days, not weeks".

@michmike, maybe you want to "file for an exception" as described in https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md ?

Since this basically trivial change is waiting for more than a year now (as successor of #80917), I fear that it it doesn't go in now into 1.120, it will be lost forever...

**Comment by michmike** [comment]
hi all, unfortunately this PR will *not* make it into v1.20. we could not get an exception for it. The good news is that now we have e2e tests that our team is working on validating and we can merge this early in v1.21. it is imperative that this PR goes in early in 1.21 to give it time to bake and stabilize versus trying to push it in the last days of the release.

my apologies to anyone affected by this.
/milestone v1.21

**Comment by thomacos** [comment]
What about merging it now?

I mean, not just in some few days, weeks, months or never. But right now. :-)
Since we are after Thaw.

**Comment by jsturtevant** [comment]
@thomacos we discussed at sig-windows and it sounds like we are ready to go ahead. Could you do a rebase?

**Comment by aarnaud** [comment]
I will rebase it this afternoon ;-)

Anthony ARNAUD
________________________________

**Comment by aarnaud** [comment]
Rebased, remove the commit that fix the `numa` when `nil` a fix was already pushed on master

**Comment by jsturtevant** [comment]
@aarnaud looks like there is still a conflict in `devicemanager/manager.go`

**Comment by aarnaud** [comment]
> @aarnaud looks like there is still a conflict in `devicemanager/manager.go`

my bad, sorry it was late, and I forgot to push. It's why the test didn't start too.

**Comment by aarnaud** [comment]
All test passed including pull-kubernetes-e2e-aks-engine-windows-gpu

**Comment by aarnaud** [comment]
@jsturtevant who can merge ? you ?

**Comment by michmike** [comment]
we need someone from sig-node to approve. Yuju, can you please approve.
@liggitt since you reviewed this in the past, can you also please approve?
/assign @yujuhong

**Comment by thomacos** [comment]
Oh no room for merging this PR huh?
Fine!
We'll go build our own Kubernetes fork.
With [blackjack](https://user-images.githubusercontent.com/54941254/102930024-f8d10680-449b-11eb-9ca9-9bff75a62333.mp4).
And [hookers](https://user-images.githubusercontent.com/54941254/102929959-da6b0b00-449b-11eb-9c7b-dff1afacf3ed.mp4).

🕯️ Happy holidays to everyone! 😀

**Comment by michmike** [comment]
> Oh no room for merging this PR huh?
> Fine!
> We'll go build our own Kubernetes fork.
> With [blackjack](https://user-images.githubusercontent.com/54941254/102930024-f8d10680-449b-11eb-9ca9-9bff75a62333.mp4).
> And [hookers](https://user-images.githubusercontent.com/54941254/102929959-da6b0b00-449b-11eb-9c7b-dff1afacf3ed.mp4).
> 
> 🕯️ Happy holidays to everyone! 😀

@thomacos with the holidays, a lot of folks are out. we are trying to get this merged ASAP

**Comment by thomacos** [comment]
> > Oh no room for merging this PR huh?
> > Fine!
> > We'll go build our own Kubernetes fork.
> > With [blackjack](https://user-images.githubusercontent.com/54941254/102930024-f8d10680-449b-11eb-9ca9-9bff75a62333.mp4).
> > And [hookers](https://user-images.githubusercontent.com/54941254/102929959-da6b0b00-449b-11eb-9c7b-dff1afacf3ed.mp4).
> > 🕯️ Happy holidays to everyone! 😀
> 
> @thomacos with the holidays, a lot of folks are out. we are trying to get this merged ASAP

No problem! Everything is fine. Was more meant as a light-hearted comment 😀.
(With a grain of truth, but definitely not directed at you sig-windows guys - thanks for all!)
Enjoy your earned free days, @michmike!

**Comment by thomacos** [comment]
@dchen1107 thanks for approving!
@michmike thanks for your telekineticapproval abilities!

**Comment by michmike** [comment]
@aarnaud and @thomacos this is an early xmas present! thank you to all that helped make this feature possible

**Comment by aarnaud** [comment]
thanks to all. Merry Christmas

**Comment by salanki** [comment]
Great work @aarnaud

**Comment by TBBle** [comment]
If this makes device plugins work on Windows (which is my understanding) at least with dockershim, that seems like a user-visible change that should be in the [v1.21 changelog](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.21.md), but is not.

Or this being deliberately sat-upon until #97739 is resolved so that non-deprecated container runtimes can support this?

**Comment by jsturtevant** [comment]
Looks like there was no release note added.   We should probably add one but I don't know the process for updating after the fact

**Comment by marosset** [comment]
> Looks like there was no release note added. We should probably add one but I don't know the process for updating after the fact

I think @aarnaud can just edit the initial PR comment to add a release note.
I agree this should be mentioned.

**Comment by aarnaud** [comment]
I just added ;-)

**Pull request merged**
